src/lib/core/datetime/date-adapter.ts

```diff
@@ -0,0 +1,196 @@
+   */
+  abstract getDayOfWeek(date: D): number;
+
+  /**
+   * Gets a list of names for the months.
+   * @param style The naming style (e.g. long = 'January', short = 'Jan', narrow = 'J').
+   * @returns An ordered list of all month names, starting with January.
+   */
+  abstract getMonthNames(style: 'long' | 'short' | 'narrow'): string[];
+
+  /**
+   * Gets a list of names for the dates of the month.
+   * @returns An ordered list of all date of the month names, starting with '1'.
+   */
+  abstract getDateNames(): string[];
+
+  /**
+   * Gets a list of names for the days of the week.
+   * @param style The naming style (e.g. long = 'Sunday', short = 'Sun', narrow = 'S').
+   * @returns An ordered list of all weekday names, starting with Sunday.
+   */
+  abstract getDayOfWeekNames(style: 'long' | 'short' | 'narrow'): string[];
+
+  /**
+   * Gets the name for the year of the given date.
+   * @param date The date to get the year name for.
+   * @returns The name of the given year (e.g. '2017').
+   */
+  abstract getYearName(date: D): string;
+
+  /**
+   * Gets the name for the month and year of the given date.
+   * @param date The date to get the month and year name for.
+   * @param monthStyle The naming style for the month
+   *     (e.g. long = 'January', short = 'Jan', narrow = 'J').
+   * @returns The name of the month and year of the given date (e.g. 'Jan 2017').
+   */
+  abstract getMonthYearName(date: D, monthStyle: 'long' | 'short' | 'narrow'): string;
+
+  /**
+   * Gets the first day of the week.
+   * @returns The first day of the week (0-indexed, 0 = Sunday).
+   */
+  abstract getFirstDayOfWeek(): number;
+
+  /**
+   * Creates a date with the given year, month, and date.
+   * @param year The full year of the date. (e.g. 89 means the year 89, not the year 1989).
+   * @param month The month of the date (0-indexed, 0 = January). If `month` is less than 0 or
+   *     greater than 11, it should roll into the previous / next year.
+   * @param date The date of month of the date. If `date` is less than 1 or greater than the number
+   *     of days in the `month`, it should roll into the previous / next month.
+   * @returns The new date.
+   */
+  abstract createDate(year: number, month: number, date: number): D;
+
+  /**
+   * Gets today's date.
+   * @returns Today's date.
+   */
+  abstract today(): D;
+
+  /**
+   * Parses a date from a value.
+   * @param value The value to parse.
+   * @param fmt The expected format of the value being parsed (type is implementation-dependent).
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract parse(value: any, fmt?: any): D | null;
+
+  /**
+   * Formats a date as a string.
+   * @param date The value to parse.
+   * @param fmt The format to use for the result string.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract format(date: D, fmt?: any): string;
```

Comment:
I think this should be ok `Intl.DateTimeFormat`, `moment.js`, and `goog.date.Date` all support this